eipw-lint/src/lints/markdown/link_eip.rs

```diff
@@ -0,0 +1,177 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use annotate_snippets::snippet::{Annotation, AnnotationType, Slice, Snippet};
+
+use comrak::nodes::{Ast, NodeLink};
+
+use crate::lints::{Context, Error, Lint};
+use crate::tree::{self, Next, TraverseExt};
+
+use regex::Regex;
+
+use serde::{Deserialize, Serialize};
+
+use std::fmt::{Debug, Display};
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct LinkEip<S>(pub S);
+
+impl<S> Lint for LinkEip<S>
+where
+    S: Display + Debug + AsRef<str>,
+{
+    fn lint<'a>(&self, slug: &'a str, ctx: &Context<'a, '_>) -> Result<(), Error> {
+        let pattern = self.0.as_ref();
+        let re = Regex::new(pattern).map_err(Error::custom)?;
+
+        let mut visitor = Visitor {
+            ctx,
+            re,
+            slug,
+            link_depth: 0,
+            text_depth: 0,
+            current_link: Link {
+                url: String::new(),
+                text: String::new(),
+            },
+        };
+        ctx.body().traverse().visit(&mut visitor)?;
+
+        Ok(())
+    }
+}
+#[derive(Debug)]
+struct Link {
+    url: String,
+    text: String,
+}
+
+#[derive(Debug)]
+struct Visitor<'a, 'b, 'c> {
+    ctx: &'c Context<'a, 'b>,
+    re: Regex,
+    slug: &'c str,
+    link_depth: usize,
+    text_depth: usize,
+    current_link: Link,
+}
+
+impl<'a, 'b, 'c> Visitor<'a, 'b, 'c> {
+    fn extract_capture(&self, text: &str, re: &Regex, index: usize) -> Result<String, Error> {
+        if let Some(captures) = re.captures(text) {
+            Ok(captures
+                .get(index)
+                .map(|m| m.as_str().to_string())
+                .unwrap_or_default())
```

Comment:
I'd say that if the regular expression has the wrong number of capture groups, we should inform the user instead of silently returning the empty string.

You can ignore my previous comment about simplifying, and use `Error::custom` and something like:

```rust
#[derive(Debug, Snafu)]
struct BadRegex;
```